File: code/data_structures/src/queue/reverse_queue/reverse_queue.go
// Part of Cosmos by OpenGenus Foundation
package main

/*
Exptected output:
Current Queue is [], Is Queue empty ? true
Try to push 30 into Queue
Try to push 12 into Queue
Try to push 34 into Queue
Try to push 2 into Queue
Try to push 17 into Queue
Try to push 88 into Queue
Reverse Queue
88 17 2 34 12 30


*/
import "fmt"

type Queue []int

func (q *Queue) Push(v int) {
	*q = append(*q, v)
}

func (q *Queue) Pop() {
	length := len(*q)
	if length != 0 {
		(*q) = (*q)[1:]
	}
}

func (q *Queue) Top() int {
	return (*q)[0]
}

func (q *Queue) Empty() bool {
	if len(*q) == 0 {
		return true
	}

	return false
}

func (q *Queue) Reverse() {
	for i, j := 0, len(*q)-1; i < j; i, j = i+1, j-1 {
		(*q)[i], (*q)[j] = (*q)[j], (*q)[i]
	}
}

func main() {
	queue := Queue{}
	fmt.Printf("Current Queue is %s, Is Queue empty ? %v \n", queue, queue.Empty())
	fmt.Printf("Try to push 30 into Queue\n")
	queue.Push(30)
	fmt.Printf("Try to push 12 into Queue\n")
	queue.Push(12)
	fmt.Printf("Try to push 34 into Queue\n")
	queue.Push(34)
	fmt.Printf("Try to push 2 into Queue\n")
	queue.Push(2)
	fmt.Printf("Try to push 17 into Queue\n")
	queue.Push(17)
	fmt.Printf("Try to push 88 into Queue\n")
	queue.Push(88)

	fmt.Println("Reverse Queue")
	queue.Reverse()
	for !queue.Empty() {
		fmt.Printf("%d ", queue.Top())
		queue.Pop()
	}
	fmt.Printf("\n")

}

File: code/data_structures/src/queue/reverse_queue/reverse_queue.js
/**
 * Part of Cosmos by OpenGenus Foundation
 */

function Queue() {
  this.queue = [];
  this.stack = [];
  this.frontEl = null;
  this.total = 0;

  this.enqueue = function (el) {
    if (this.total == 0) {
      this.frontEl = el;
    }

    this.queue.push(el);
    this.total++;
  };

  this.dequeue = function () {
    // Confirm if stack is empty and return
    if (this.total == 0) {
      console.log("The queue is empty!");

      return;
    }

    // Move all queue entries to stack if stack is empty
    if (this.stack.length == 0) {
      while (this.queue.length > 0) {
        this.stack.push(this.queue.pop());
      }
    }

    this.total--;
    // Return the topmost element
    return this.stack.pop();
  };

  this.size = function () {
    return this.total;
  };

  this.front = function () {
    if (this.stack.length > 0) {
      return this.stack[this.stack.length - 1];
    }

    return this.frontEl;
  };

  this.reverse = function () {
    if (this.total == 0) {
      console.log("The queue is empty");

      return;
    }

    this.stack.reverse();
    this.queue.reverse();
    while (this.queue.length > 0) {
      this.stack.push(this.queue.pop());
    }

    this.frontEl = this.stack[this.stack.length - 1];
  };

  this.print = function () {
    if (this.total == 0) {
      console.log(this.queue);
      return;
    }

    let output = [];
    for (let i = this.stack.length - 1; i >= 0; --i) {
      output.push(this.stack[i]);
    }
    for (let q of this.queue) {
      output.push(q);
    }

    console.log(output);
  };
}

const queue = new Queue();
const items = [1, 2, 3, 4];
for (let i of items) {
  queue.enqueue(i);
}
console.log("\nCurrent queue");
console.log("===============");
queue.print();

console.log("\nCurrent queue front");
console.log("===============");
console.log(queue.front());

const el = queue.dequeue();
console.log("\nRemoved element:");
console.log("===============");
console.log(el);

console.log("\nCurrent queue");
console.log("===============");
queue.print();

queue.enqueue(5);
queue.enqueue(6);
console.log("\nUpdated queue");
console.log("===============");
queue.print();

console.log("\nReversed queue");
console.log("===============");
queue.reverse();
queue.print();

console.log("\nReversed queue front");
console.log("===============");
console.log(queue.front());
